### ISSUE TITLE
Problem with importing the "reverb" package with Tutorial: SAC minitaur with the Actor-Learner API

### ISSUE THREAD

**Issue by orshemtov** [tensorflow/agents#532]
HI,

I am getting an ImportError when trying to import the "reverb" package as done in the tutorial.

```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-38745e83da94> in <module>
      4 import matplotlib.pyplot as plt
      5 import os
----> 6 import reverb
      7 import tempfile
      8 import PIL.Image

~/Desktop/AI/ai_venv/lib/python3.7/site-packages/reverb/__init__.py in <module>
     25 # pylint: enable=g-bad-import-order
     26 
---> 27 from reverb import item_selectors as selectors
     28 from reverb import rate_limiters
     29 

~/Desktop/AI/ai_venv/lib/python3.7/site-packages/reverb/item_selectors.py in <module>
     17 import functools
     18 
---> 19 from reverb import pybind
     20 
     21 Fifo = pybind.FifoSelector

~/Desktop/AI/ai_venv/lib/python3.7/site-packages/reverb/pybind.py in <module>
----> 1 import tensorflow as _tf; from .libpybind import *; del _tf

ImportError: libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
```

I have tried to export this variable:
export LD_LIBRARY_PATH=/home/orie/Desktop/AI/ai_venv/lib/

I have also tried including this environment variable in my python notebook:
```
import os
os.environ['LD_LIBRARY_PATH'] = '/home/orie/Desktop/AI/ai_venv/lib/'
```

I also tried:
`sudo ldconfig /home/orie/Desktop/AI/ai_venv/lib
`
I'm using Ubuntu and a virtual environment.

Thx for anyone who helps!

**Comment by orshemtov** [comment]
Even when I try to run this in the Colab notebook, trying to import reverb is causing this error:

`NotFoundError: /usr/local/lib/python3.6/dist-packages/reverb/cc/ops/libgen_dataset_op_gen_op.so: undefined symbol: _ZN10tensorflow4data15DatasetOpKernel11TraceStringEPNS_15OpKernelContextEb`

This is from this notebook:
https://colab.research.google.com/github/tensorflow/agents/blob/master/docs/tutorials/7_SAC_minitaur_tutorial.ipynb#scrollTo=sMitx5qSgJk1

**Comment by coleridge72** [comment]
I also had this issue starting from yesterday. I think it is related to the new tensorflow 2.4

It seems to be working if you install tf 2.3.1 just before installing reverb & tf agents in that colab notebook
```
!pip install tensorflow==2.3.1
!pip install tf-agents[reverb]
```

**Comment by jeffheaton** [comment]
Also getting this error.

**Comment by MTDzi** [comment]
Same here.

**Comment by erikbixby** [comment]
Same.

**Comment by tfboyd** [comment]
All fixed with TF-Agents release 0.7.1. I tested the colab in tensorflow.org/agents.  I am looking at locking TF-Agents to a specific minor release of TensorFlow to avoid this in the future but only if installing with `[reverb]`.  This will still allow people to try to mix and match TF versions or custom builds but use the `tf-agents[reverb]` directive as the most tested path.  Sorry about that, TF got released just before the holiday.

**Comment by Enzodtz** [comment]
Hey, maybe too late but also got his error. I've fixed it with `pip install reverb`, but as @tfboyd said, this maybe isn't the best solution.

**Comment by Nik-Kras** [comment]
Hi, everyone. I am trying to follow the tutorial - https://www.tensorflow.org/agents/tutorials/1_dqn_tutorial
However, I face errors in importing phase. 

`Traceback (most recent call last):

  File "D:\-----\main.py", line 37, in <module>
    import reverb
  File "D:\-----\venv\lib\site-packages\reverb.py", line 64
    raise TypeError, 'Regexp cannot be negated'
                   ^
SyntaxError: invalid syntax`

I tried different combinations of package versions. Tried to install the last versions from tutorials (not only official ones). Currently, the configurations are next:

`
Keras-Preprocessing | 1.1.2 | 1.1.2 

Markdown | 3.4.1 | 3.4.1
MarkupSafe | 2.1.1 | 2.1.1
Pillow | 9.2.0 | 9.2.0
PyVirtualDisplay | 3.0 | 3.0
Pygments | 2.13.0 | 2.13.0
Werkzeug | 2.2.2 | 2.2.2
absl-py | 1.2.0 | 1.2.0
asgiref | 3.5.0 | 3.5.2
asttokens | 2.0.8 | 2.0.8
astunparse | 1.6.3 | 1.6.3
backcall | 0.2.0 | 0.2.0
cachetools | 5.2.0 | 5.2.0
certifi | 2022.6.15 | 2022.6.15
charset-normalizer | 2.1.1 | 3.0.0b2
cloudpickle | 2.1.0 | 2.1.0
colorama | 0.4.5 | 0.4.5
cycler | 0.11.0 | 0.11.0
decorator | 5.1.1 | 5.1.1
dm-tree | 0.1.7 | 0.1.7
executing | 0.10.0 | 0.10.0
flatbuffers | 1.12 | 2.0.7
fonttools | 4.37.1 | 4.37.1
gast | 0.4.0 | 0.5.3
gin-config | 0.1.3 | 0.5.0
google-auth | 2.11.0 | 2.11.0
google-auth-oauthlib | 0.4.6 | 0.5.2
google-pasta | 0.2.0 | 0.2.0
grpcio | 1.47.0 | 1.49.0rc1
gym | 0.23.0 | 0.25.2
gym-notices | 0.0.8 | 0.0.8
h5py | 3.7.0 | 3.7.0
idna | 3.3 | 3.3
imageio | 2.4.0 | 2.21.1
ipython | 8.4.0 | 8.4.0
jedi | 0.18.1 | 0.18.1
keras | 2.9.0 | 2.10.0rc1
kiwisolver | 1.4.4 | 1.4.4
libclang | 14.0.6 | 14.0.6
matplotlib | 3.5.3 | 3.6.0rc1
matplotlib-inline | 0.1.6 | 0.1.6
numpy | 1.23.2 | 1.23.2
oauthlib | 3.2.0 | 3.2.0
opt-einsum | 3.3.0 | 3.3.0
packaging | 21.3 | 21.3
parso | 0.8.3 | 0.8.3
pickleshare | 0.7.5 | 0.7.5
pip | 22.2.2 | 22.2.2
prompt-toolkit | 3.0.30 | 3.0.30
protobuf | 3.19.4 | 4.21.5
pure-eval | 0.2.2 | 0.2.2
pyasn1 | 0.4.8 | 0.5.0rc2
pyasn1-modules | 0.2.8 | 0.3.0rc1
pygame | 2.1.0 | 2.1.3.dev4
pyglet | 1.5.26 | 2.0a4
pyparsing | 3.0.9 | 3.0.9
python-dateutil | 2.8.2 | 2.8.2
requests | 2.28.1 | 2.28.1
requests-oauthlib | 1.3.1 | 1.3.1
reverb | 2.0.1 | 2.0.1
rsa | 4.9 | 4.9
setuptools | 63.2.0 | 65.3.0
six | 1.16.0 | 1.16.0
sqlparse | 0.4.2 | 0.4.2
stack-data | 0.4.0 | 0.4.0
tensorboard | 2.9.1 | 2.10.0
tensorboard-data-server | 0.6.1 | 0.6.1
tensorboard-plugin-wit | 1.8.1 | 1.8.1
tensorflow | 2.9.1 | 2.10.0rc2
tensorflow-estimator | 2.9.0 | 2.10.0rc0
tensorflow-io-gcs-filesystem | 0.26.0 | 0.26.0
tensorflow-probability | 0.17.0 | 0.17.0
termcolor | 1.1.0 | 1.1.0
tf-agents | 0.5.0 | 0.14.0rc0
traitlets | 5.3.0 | 5.3.0
typing-extensions | 4.3.0 | 4.3.0
tzdata | 2022.1 | 2022.2
urllib3 | 1.26.12 | 1.26.12
wcwidth | 0.2.5 | 0.2.5
wheel | 0.37.1 | 0.37.1
wrapt | 1.14.1 | 1.14.1
`

I am working on Windows OS, using python3.

**Comment by hopskipnfall** [comment]
According to the Reverb [README](https://github.com/deepmind/reverb#installation), "Reverb currently only supports Linux based OSes."

So if you're on Windows it's not going to work :/ 

See related issue #695

**Comment by Nik-Kras** [comment]
> According to the Reverb [README](https://github.com/deepmind/reverb#installation), "Reverb currently only supports Linux based OSes."
> 
> So if you're on Windows it's not going to work :/
> 
> See related issue #695

Thank you for your reply
This tutorial from your link (https://github.com/tensorflow/agents/blob/master/docs/tutorials/9_c51_tutorial.ipynb) doesn't use reverb and works well!

But, shouldn't that dependency on OS be fixed in the tutorial? I may be mistaken as I don't have much experience, but for me, it looks like an issue that should be solved by replacing the reverb package with something more cross-platform. Therefore, the basic tutorial would work in Windows as it supposed to be

**Comment by hopskipnfall** [comment]
I agree.. I hope the maintainers update the tutorial to at least mention that it is not supported on Windows.

**Comment by tfboyd** [comment]
Sorry about that, I did not think to update the tf/agents pages with the
linux situation created by Reverb.  TF-Agents previously was all python and
didn't work; but we needed to move to Reverb so we did not have to support
our own replay buffer and Reverb is also more performance than our past
solution.  I will try to remember to add a note to the github index, I am
not sure I will add it to all the notebook files; but I will not rule it
out in trying to make sure it is visible.  Sorry about that, that would
have made me frustrated as well.

**Comment by hopskipnfall** [comment]
Thanks Toby!

**Comment by tfboyd** [comment]
PR is in review.  Not the best fix, but I added a warning to the Install section of the front page of github.com.  Off-the-cuff adding it to all the tutorials feels messy.  Open to other ideas. I know it would have been great for you to see it right there in the notebook.

**Comment by FinAminToastCrunch** [comment]
The fact that Reverb is not OS agnostic is a massive problem. It at least needed to be mentioned in the official docs. A vast majority of the tutorials use Reverb. I understand that it is good for a performance standpoint, but a cross platform alternative needs to be found.

**Comment by tfboyd** [comment]
This is much more than you want to know; but insight into the people behind the code is often a useful perspective. 

@FinAminToastCrunch  I added a note to the README.md in the install section a few days ago.  Reverb (I realize you would not visit that site) also has a big note about only supporting Linux. Having a note on all of the tutorials feels messy. We only use Linux or linux like systems internally. There was an effort to compile reverb (c++) on OSX; but it didn't happen and there isn't an incentive to support it internally. The code is there and anyone is welcome to figure it out; bazel is a bit of a pain which is a rather large barrier to entry.  We would still want to setup internal regression testing and presubmits for OSX (Which is kind of a pain).

TF-Agents use to have a python based replay buffer; but we decided to move to Reverb which aligns with ACME and other RL frameworks across Alphabet, e.g. DeepMind. It is an economies of scale situation (we no longer needed to have an engineer supporting and extending the replay buffer) and it meant everyone (across "all" of Alphabet) focused on adding features and performance to the same code base. 

I do not mean this in a flippant manner. If your requirement is to run on Windows and OSX, then TF-Agents is not the right choice unless you are just using it for the model code and coming up with a different replay buffer. Reverb as a distributed replay buffer has been testing with 1000s of agents; which doesn't help your situation but is why we moved.

I honestly wish you luck in finding the right tool for the work you want to accomplish.  I didn't have to respond; and while this doesn't help you. I hope it gives you some perspective.

**Comment by FinAminToastCrunch** [comment]
Thanks for the perspective. I'm happy to see the decision to move to reverb has saved you and your teammates lots of time.

**Comment by doughazell** [comment]
> > According to the Reverb [README](https://github.com/deepmind/reverb#installation), "Reverb currently only supports Linux based OSes."
> > So if you're on Windows it's not going to work :/
> > See related issue #695
> 
> Thank you for your reply This tutorial from your link (https://github.com/tensorflow/agents/blob/master/docs/tutorials/9_c51_tutorial.ipynb) doesn't use reverb and works well!
> 
> But, shouldn't that dependency on OS be fixed in the tutorial? I may be mistaken as I don't have much experience, but for me, it looks like an issue that should be solved by replacing the reverb package with something more cross-platform. Therefore, the basic tutorial would work in Windows as it supposed to be

Thanks this really helped me to develop https://github.com/doughazell/mnist-training-errors which is Keras Sequential to GSheet and CategoricalQNetwork on MacOS

**Comment by Nik-Kras** [comment]
Great
I am glad it worked!

**Comment by EkaterinaAbramova** [comment]
So just to be clear, since I am on Mac I cannot use TF Agents? So what would you recommend to us (Mac, windows users)? Is there another buffer that we can substitute / especially for beginners with TF Agents (I don't know how to write my own). What's a work around / other way of running it? My current thinking is to try using AWS linux machine?

**Comment by doughazell** [comment]
@EkaterinaAbramova you need to use TFUniformReplayBuffer (and not ReverbReplayBuffer that only works on Linux) on Mac like I do in:

https://github.com/doughazell/mnist-training-errors/blob/a2d02ea98562a5f1c8bbcf08b6dddcaba49faf0b/suite_gym_utils.py#L81

**Comment by tfboyd** [comment]
@EkaterinaAbramova  You are correct.
@doughazell  is also correct; but I will say we do not use those ReplayBuffers anymore. Meaning they do not get a lot of exercise; but that is what we used up until about 1.5 years ago or so when we switched. They are likely just fine for playing around and if they help you go for it.

I would use an AWS machine. I have not tried docker on Mac, but that might be an option. I only read this 3 minutes ago but it seems it is free if are are not using it commercially.  https://docs.docker.com/desktop/install/mac-install/

The long answer is we use a replay buffer called reverb that is in c++ and we (DeepMind now Google Deepmind "owns" it; but I do the releases for them and we are very close) do not have enough people to manage a MacOS build. Internally we are linux and that is the best path for support. I would not try to write you own, I would run a linux on something.
